static/js/components/dashboard/courses/StatusMessages.js

```diff
@@ -0,0 +1,188 @@
+// @flow
+import React from 'react';
+import R from 'ramda';
+import moment from 'moment';
+
+import type { Coupon } from '../../../flow/couponTypes';
+import type { Course, CourseRun } from '../../../flow/programTypes';
+import { couponMessageText } from '../CouponMessage';
+import { makeCouponReason } from '../../../lib/coupon';
+import {
+  COUPON_CONTENT_TYPE_COURSE,
+  COURSE_ACTION_PAY,
+  COURSE_ACTION_REENROLL,
+} from '../../../constants';
+import { S } from '../../../lib/sanctuary';
+import {
+  hasPaidForAnyCourseRun,
+  courseUpcomingOrCurrent,
+  hasPassedCourseRun,
+  hasFailedCourseRun,
+  futureEnrollableRun,
+} from './util';
+import { hasPassingExamGrade } from '../../../lib/grades';
+
+type Message = {
+  message: string|React$Element<*>,
+  action?: React$Element<*>
+};
+
+export const formatAction = (message: Message) => (
+  <div className="second-col">
+    { message.action }
+  </div>
+);
+
+export const formatMessage = (message: Message, index?: number) => (
+  <div className="status-message cols" key={index}>
+    <div className="message first-col">
+      { message.message }
+    </div>
+    { message.action ? formatAction(message) : null }
+  </div>
+);
+
+const formatDate = date => (
+  moment(date).format('MM/DD/YYYY')
+);
+
+type CalculateMessagesProps = {
+  courseAction:                (run: CourseRun, actionType: string) => React$Element<*>,
+  firstRun:                    CourseRun,
+  course:                      Course,
+  expandedStatuses:            Set<number>,
+  setShowExpandedCourseStatus: (n: number) => void,
+  coupon?:                     Coupon,
+};
+
+// this calculates any status messages we'll need to show the user
+// we wrap the array of messages in a Maybe, so that we can indicate
+// the case where there are no messages cleanly
+// 
+// I'm so sorry for anyone who has to read / modify this
+// sometimes there really isn't a better way :/
+export const calculateMessages = (props: CalculateMessagesProps) => {
+  const {
+    coupon,
+    courseAction,
+    firstRun,
+    course,
+    expandedStatuses,
+    setShowExpandedCourseStatus,
+  } = props;
+
+  let exams = course.has_exam;
+  let paid = firstRun.has_paid;
+  let passedExam = hasPassingExamGrade(course);
+  let paymentDueDate = moment(firstRun.course_upgrade_deadline);
+
+  // Course is running, user has already paid, we show no messages
+  if (courseUpcomingOrCurrent(firstRun) && paid) {
+    return S.Nothing; 
+  }
+
+  let messages = [];
+
+
+  if (coupon && coupon.content_type === COUPON_CONTENT_TYPE_COURSE && coupon.object_id === course.id) {
+    messages.push({ message: `${couponMessageText(coupon)}${makeCouponReason(coupon)}.`});
```

Comment:
Can you refactor the concatenation here into its own function? It looks like `CouponMessage` does a similar thing, except that it wraps it in a `<div className='coupon-message' />`. It would be good to share code between the two, and the coupon message concatenation is a bit brittle IMO so it would be good to abstract away from it